  Add Substratum dashboard tile

Modified to use parts of @tsubus launching logic

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index 1210070..0089013 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -2845,9 +2845,29 @@
                 android:value="true" />
         </activity>
 
-        <activity android:name=".applications.LockPatternActivity"
-                  android:label="@string/protected_apps"
-                  android:excludeFromRecents="true">
+        <activity android:name="Settings$SubstratumActivity"
+            android:label="@string/substratum_title"
+            android:icon="@drawable/ic_settings_substratum"
+            android:taskAffinity="">
+            <intent-filter>
+                <action android:name="android.intent.action.MAIN" />
+                <category android:name="android.intent.category.DEFAULT" />
+                <category android:name="com.android.settings.SHORTCUT" />
+            </intent-filter>
+            <intent-filter android:priority="0">
+                <action android:name="com.android.settings.action.SETTINGS" />
+            </intent-filter>
+            <meta-data android:name="com.android.settings.category"
+                android:value="com.android.settings.category.system" />
+            <meta-data android:name="com.android.settings.FRAGMENT_CLASS"
+                android:value="com.android.settings.Substratum" />
+            <meta-data android:name="com.android.settings.summary"
+                android:resource="@string/substratum_summary" />
+        </activity>
+
+        <activity android:name=".applications.LockPatternActivity"
+            android:label="@string/protected_apps"
+            android:excludeFromRecents="true">
             <intent-filter>
                 <action android:name=".applications.LockPatternActivity" />
                 <category android:name="android.intent.category.DEFAULT" />
diff --git a/src/com/android/settings/Settings.java b/src/com/android/settings/Settings.java
index 1310070..0189013 100644
--- a/src/com/android/settings/Settings.java
+++ b/src/com/android/settings/Settings.java
@@ -72,6 +72,7 @@ public boolean isValidFragment(String className) {
     public static class StorageUseActivity extends SettingsActivity { /* empty */ }
     public static class DevelopmentSettingsActivity extends SettingsActivity { /* empty */ }
     public static class SuperSUActivity extends SettingsActivity { /* empty */ }
+    public static class SubstratumActivity extends SettingsActivity { /* empty */ }
     public static class AccessibilitySettingsActivity extends SettingsActivity { /* empty */ }
     public static class CaptioningSettingsActivity extends SettingsActivity { /* empty */ }
     public static class AccessibilityInversionSettingsActivity extends SettingsActivity { /* empty */ }
diff --git a/res/values/projekt_strings.xml b/res/values/projekt_strings.xml
index 1110070..0289013 100644
--- a/res/values/projekt_strings.xml
+++ b/res/values/projekt_strings.xml
@@ -29,4 +29,8 @@
    <string name="force_authorize_substratum_packages_title">Force authorize every theme app</string>
    <string name="force_authorize_substratum_packages_summary">Allow theme apps from unknown sources</string>
 
+   <!-- Top Level Substratum settings -->
+   <string name="substratum_title">Substratum</string>
+   <string name="substratum_summary">Customize device theme</string>
+
 </resources>
diff --git a/res/values-ru/projekt_strings.xml b/res/values-ru/projekt_strings.xml
index 1110070..0389013 100644
--- a/res/values-ru/projekt_strings.xml
+++ b/res/values-ru/projekt_strings.xml
@@ -29,4 +29,8 @@
    <string name="force_authorize_substratum_packages_title">Разрешить установку тем из неизвестных источников</string>
    <string name="force_authorize_substratum_packages_summary">Принудительно запрашивать разрешения для каждой темы</string>
 
+   <!-- Top Level Substratum settings -->
+   <string name="substratum_title">Темы Substratum</string>
+   <string name="substratum_summary">Установка тем оформления</string>
+
 </resources>
diff --git a/src/com/android/settings/SettingsActivity.java b/src/com/android/settings/SettingsActivity.java
index 1110070..0489015 100644
--- a/src/com/android/settings/SettingsActivity.java
+++ b/src/com/android/settings/SettingsActivity.java
@@ -243,6 +243,8 @@
 
     private static final String SUPERSU_FRAGMENT = "com.android.settings.SuperSU";
 
+    private static final String SUBSTRATUM_FRAGMENT = "com.android.settings.Substratum";
+
     private String mFragmentClass;
     private String mActivityAction;
 
@@ -1099,7 +1101,14 @@ public void startPreferenceFragment(Fragment fragment, boolean push) {
             startActivity(superSUIntent);
             finish();
             return null;
+        } else if (SUBSTRATUM_FRAGMENT.equals(fragmentName)) {
+            Intent subIntent = new Intent();
+            subIntent.setClassName("projekt.substratum", "projekt.substratum.LaunchActivity");
+            startActivity(subIntent);
+            finish();
+            return null;
         }
+
         if (validate && !isValidFragment(fragmentName)) {
             throw new IllegalArgumentException("Invalid fragment for this activity: "
                     + fragmentName);
@@ -1206,6 +1215,16 @@ private void doUpdateTilesList() {
         // Reveal development-only quick settings tiles
         DevelopmentTiles.setTilesEnabled(this, showDev);
 
+        // Substratum
+        boolean subSupported = false;
+        try {
+            subSupported = (getPackageManager().getPackageInfo("projekt.substratum", 0).versionCode > 0);
+        } catch (PackageManager.NameNotFoundException e) {
+        }
+        setTileEnabled(new ComponentName(packageName,
+                        Settings.SubstratumActivity.class.getName()),
+                subSupported, isAdmin, pm);
+
         // Show scheduled power on and off if support
         boolean showTimerSwitch = false;
         Intent intent = new Intent(ACTION_TIMER_SWITCH);
diff --git a/res/drawable/ic_settings_substratum.xml b/res/drawable/ic_settings_substratum.xml
new file mode 100644
index 111f231..0589000
--- /dev/null
+++ b/res/drawable/ic_settings_substratum.xml
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+    Copyright (c) 2016 Project Substratum
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+-->
+
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+        android:width="24dp"
+        android:height="24dp"
+        android:viewportHeight="24"
+        android:viewportWidth="24">
+    <path
+        android:fillColor="?android:attr/colorAccent"
+        android:pathData="M20.71,4.63L19.37,3.29C19,2.9 18.35,2.9 17.96,3.29L9,12.25L11.75,15L20.71,6.04C21.1,5.65 21.1,5 20.71,4.63M7,14A3,3 0 0,0 4,17C4,18.31 2.84,19 2,19C2.92,20.22 4.5,21 6,21A4,4 0 0,0 10,17A3,3 0 0,0 7,14Z"/>
+</vector>
diff --git a/res/drawable/ic_settings_substratum.xml b/res/drawable/ic_settings_substratum.xml
index 1110070..0689013 100644
--- a/res/drawable/ic_settings_substratum.xml
+++ b/res/drawable/ic_settings_substratum.xml
@@ -1,26 +1,49 @@
 <?xml version="1.0" encoding="utf-8"?>
-<!--
-    Copyright (c) 2016 Project Substratum
-
-    Licensed under the Apache License, Version 2.0 (the "License");
-    you may not use this file except in compliance with the License.
-    You may obtain a copy of the License at
-
-        http://www.apache.org/licenses/LICENSE-2.0
-
-    Unless required by applicable law or agreed to in writing, software
-    distributed under the License is distributed on an "AS IS" BASIS,
-    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-    See the License for the specific language governing permissions and
-    limitations under the License.
--->
-
 <vector xmlns:android="http://schemas.android.com/apk/res/android"
-        android:width="24dp"
-        android:height="24dp"
-        android:viewportHeight="24"
-        android:viewportWidth="24">
-    <path
-        android:fillColor="?android:attr/colorAccent"
-        android:pathData="M20.71,4.63L19.37,3.29C19,2.9 18.35,2.9 17.96,3.29L9,12.25L11.75,15L20.71,6.04C21.1,5.65 21.1,5 20.71,4.63M7,14A3,3 0 0,0 4,17C4,18.31 2.84,19 2,19C2.92,20.22 4.5,21 6,21A4,4 0 0,0 10,17A3,3 0 0,0 7,14Z"/>
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="512.000000"
+    android:viewportHeight="512.000000"
+    android:tint="?android:attr/colorAccent">
+
+    <group
+        android:translateY="512.000000"
+        android:scaleX="0.100000"
+        android:scaleY="-0.100000">
+        <path
+            android:fillColor="#FFFFFFFF"
+            android:strokeWidth="1"
+            android:pathData="M2178 4731 c-107 -35 -180 -98 -247 -218 l-41 -72 1 -268 c0 -181 3 -263 10 -253 6
+8 73 130 150 270 82 149 151 263 166 274 14 10 45 21 69 24 41 4 66 -8 526 -258
+266 -145 495 -274 510 -288 34 -31 52 -82 43 -123 -4 -17 -42 -95 -85 -173 -42 -77
+-373 -682 -734 -1344 l-656 -1202 2 -266 3 -266 201 368 c111 203 493 902 849 1554
+356 652 653 1201 660 1221 51 144 7 321 -105 422 -52 47 -998 569 -1085 599 -64 22
+-165 22 -237 -1z" />
+        <path
+            android:fillColor="#FFFFFFFF"
+            android:strokeWidth="1"
+            android:pathData="M3672 3693 c-31 -4 -202 -87 -306 -149 -47 -28 -55 -40 -163 -238 -62 -115 -112
+-210 -111 -211 2 -1 140 75 308 169 265 148 310 171 347 170 84 -2 84 -2 381 -531
+l273 -488 -4 -54 c-3 -47 -9 -59 -38 -85 -19 -17 -582 -337 -1251 -711 l-1217 -680
+-1 -149 c0 -115 3 -147 13 -143 6 3 590 329 1297 725 923 516 1300 733 1337 766 78
+70 116 154 121 266 3 65 0 101 -13 140 -9 28 -141 273 -292 543 -306 547 -323 571
+-444 627 -68 31 -151 43 -237 33z" />
+        <path
+            android:fillColor="#FFFFFFFF"
+            android:strokeWidth="1"
+            android:pathData="M4108 1923 l-236 -128 322 -5 c356 -6 355 -5 396 -73 19 -31 20 -55 20 -597 0 -546
+-1 -566 -20 -598 -11 -18 -34 -41 -52 -52 -33 -20 -55 -20 -1960 -20 l-1928 0 0
+-130 0 -130 1958 2 1958 3 76 37 c91 45 147 103 191 196 l32 67 0 625 0 625 -32 67
+c-43 92 -100 151 -190 195 -70 35 -81 37 -186 40 l-112 3 -237 -127z" />
+        <path
+            android:fillColor="#FFFFFFFF"
+            android:strokeWidth="1"
+            android:pathData="M446 4899 c-62 -15 -153 -68 -197 -116 -22 -24 -55 -74 -72 -111 l-32 -67 0 -2061
+0 -2060 37 -76 c45 -91 103 -147 196 -191 l67 -32 625 0 625 0 67 32 c93 44 151
+100 196 191 l37 76 0 2061 0 2061 -37 76 c-45 91 -103 147 -196 191 l-67 32 -605 2
+c-345 1 -622 -3 -644 -8z m789 -3438 c81 -37 139 -96 178 -179 28 -59 32 -77 32
+-152 0 -75 -4 -93 -32 -152 -39 -84 -97 -142 -181 -181 -59 -28 -77 -32 -152 -32
+-75 0 -93 4 -152 32 -84 39 -142 97 -181 181 -28 59 -32 77 -32 152 0 75 4 93 32
+152 49 105 140 180 250 208 66 16 168 4 238 -29z" />
+    </group>
 </vector>
